Cg/lab2 implementation

Ppasha9 opened the pull request.

pv6 committed:
Add basic post-proccessing

pv6 committed:
Calculate frame average brightness

pv6 committed:
Pass adapted frame average brightness to HDR shader

pv6 committed:
Add HDR implementation

Ppasha9 committed:
Add camera movement. Start lights implementation

Ppasha9 committed:
Change calculating result pixel color in pixel shader

Ppasha9 committed:
Add three point lights with strength regulation

anthark commented on anim/LogCopyTexturePixelShader.hlsl:
```diff
@@ -0,0 +1,15 @@
+float4 main(PixelShaderInput input) : SV_TARGET
+{
+    float4 p = shaderTexture.Sample(samplerState, input.texcoord);
+    return log(p) + 1;
```
Хм... Какое число получится, если p == float4(0,0,0,0) ?

pv6 replied:
Исправил на log(p + 1). Эта ошибка взялась со слайда про уменьшающиеся текстуры (сл. 37), там написана неправильная формула.

anthark replied:
Спасибо. Поправлю. То-то я смотрю, у всех одна и та же ошибка =)

anthark reviewed (changes requested):
В целом хорошо. Даже управление удобное =)

Но, все-таки, есть замечания. 

1) По сцене не видно, что алгоритм реально работает. Все источники света имеют малую интенсивность и не пересекаются друг с другом. В итоге, нигде нет проблемы, которую, собственно, HDR решает. Сделайте с этим что-нибудь, пожалуйста. 

2) Большое количество warning'ов и error'ов. Исправьте, пожалуйста. Можно Включить break на них, тогда приложение будет сразу валится в нужных местах. 

3) При выходе остается много неудаленных объектов. Что особенно удивительно с учетом того, что вы используете автоуказатели. Исправьте, пожалуйста.

pv6 committed:
Fix problems

- fix Directx errors and warnings
- fix logarithm for zero brightness
- change cube to plane
- change light controls to keys '1', '2', '3'
- add perlin noise to the plane

pv6 committed:
Release Direct3D live objects

pv6 committed:
Release all Live Objects for real this time

pv6 commented:
Исправил замечания.

Переключение яркости источников теперь на цифры 1, 2, 3.
